packages/unified-selection/src/unified-selection/SelectionScope.ts

```diff
@@ -3,277 +3,217 @@
+      ),
+      Non3dElementNearestFunctionalElements (OriginalECInstanceId, ECInstanceId, ECClassId) AS (
+        SELECT OriginalECInstanceId, ECInstanceId, ECClassId
+        FROM Non3dElementsOrNearestFunctionalElements
+        WHERE ECClassId IS (Functional.FunctionalElement)
+      ),
+      Non3dElementsWithoutFunctionalElement (ECInstanceId, ECClassId) AS (
+        SELECT n3ewfe.OriginalECInstanceId, OriginalECClassId
+        FROM Non3dElementsOrNearestFunctionalElements n3ewfe
+          LEFT JOIN Non3dElementNearestFunctionalElements n3enfe ON n3enfe.OriginalECInstanceId = n3ewfe.OriginalECInstanceId
+        WHERE n3ewfe.ParentId IS NULL AND n3enfe.ECInstanceId IS NULL
+      ),
+      Non3dElements (ECInstanceId, ECClassId) AS (
+        SELECT ECInstanceId, ECClassId FROM Non3dElementNearestFunctionalElements
+        UNION
+        SELECT ECInstanceId, ECClassId FROM Non3dElementsWithoutFunctionalElement
+      ),
+      Non3dElementAncestorElements (ECInstanceId, ECClassId, Depth, ParentId) AS (
+        SELECT e.ECInstanceId, e.ECClassId, ${formAncestorLevelBinding(ancestorLevel, bindings)}, e.Parent.Id
+        FROM BisCore.Element e
+          JOIN Non3dElements n3e ON n3e.ECInstanceId = e.ECInstanceId
+        UNION ALL
+        SELECT pe.ECInstanceId, pe.ECClassId, n3eae.Depth - 1, pe.Parent.Id
+        FROM Non3dElementAncestorElements n3eae
+          JOIN BisCore.Element pe ON pe.ECInstanceId = n3eae.ParentId
+        WHERE ${recurseUntilRoot ? "" : "Depth > 0 AND"} ParentId IS NOT NULL
+      ),
+      Element3dAncestorElements (ECInstanceId, ECClassId, Depth, ParentId) AS (
+        SELECT ge.ECInstanceId, ge.ECClassId, ${formAncestorLevelBinding(ancestorLevel, bindings)}, ge.Parent.Id
+        FROM BisCore.GeometricElement3d ge
+        WHERE ${formIdBindings("ge.ECInstanceId", ids, bindings)}
+        UNION ALL
+        SELECT pe.ECInstanceId, pe.ECClassId, e3ae.Depth - 1, pe.Parent.Id
+        FROM Element3dAncestorElements e3ae
+          JOIN BisCore.Element pe ON pe.ECInstanceId = e3ae.ParentId
+        WHERE ${recurseUntilRoot ? "" : "Depth > 0 AND"} ParentId IS NOT NULL
+      ),
+      Element3dAncestorRelatedFunctionalElement (ECInstanceId, ClassName) AS (
+        SELECT
+          COALESCE(peff.TargetECInstanceId, e3ae.ECInstanceId),
+          ec_classname(COALESCE(peff.TargetECClassId, e3ae.ECClassId), 's.c')
+        FROM Element3dAncestorElements e3ae
+          LEFT JOIN Functional.PhysicalElementFulfillsFunction peff ON peff.SourceECInstanceId = e3ae.ECInstanceId
+        WHERE ${recurseUntilRoot ? "" : "e3ae.Depth = 0 OR"} e3ae.ParentId IS NULL
+      )
+      SELECT DISTINCT ECInstanceId, ec_classname(ECClassId, 's.c') AS ClassName
+      FROM Non3dElementAncestorElements
+      WHERE ${recurseUntilRoot ? "" : "Depth = 0 OR"} ParentId IS NULL
+      UNION ALL
+      SELECT DISTINCT ECInstanceId, ClassName FROM Element3dAncestorRelatedFunctionalElement;`;
```

Comment:
Don't need a semi-colon at the end
```suggestion
      SELECT DISTINCT ECInstanceId, ClassName FROM Element3dAncestorRelatedFunctionalElement`;
```

Reply:
Removed.